File: algorithms/graphs/bellman_ford/python/bellman_ford.py
"""
-----------------------------------------------------------------------------
Nom du fichier : bellman_ford.py
Auteur        : CAPELLE Gabin (https://github.com/CapelleGab)
Créé le      : 10-06-2025 20:39
Dernière MAJ : 10-06-2025 20:39

Description   : Implémentation de l'algorithme de Bellman-Ford en Python.
-----------------------------------------------------------------------------
"""

def bellman_ford(graph, start):
    distances = {vertex: float('inf') for vertex in graph}
    distances[start] = 0

    for _ in range(len(graph) - 1):
        for vertex in graph:
            for neighbor, weight in graph[vertex].items():
                if distances[vertex] + weight < distances[neighbor]:
                    distances[neighbor] = distances[vertex] + weight

    return distances

## Test de la fonction bellman_ford
def test_bellman_ford():
    # Exemple de graphe
    graph = {
        'A': {'B': 1, 'C': 4},
        'B': {'C': 2, 'D': 5},
        'C': {'D': 1},
        'D': {}
    }
    
    # Test 1: Vérifier les distances minimales à partir du sommet 'A'
    distances = bellman_ford(graph, 'A')
    assert distances == {'A': 0, 'B': 1, 'C': 3, 'D': 4}, "❌ Test échoué: distances incorrectes"
    print("✅ Test 1 réussi")

    # Test 2: Vérifier le comportement avec un graphe contenant des arêtes négatives
    graph_negative = {
        'A': {'B': 1, 'C': 4},
        'B': {'C': -3, 'D': 2},
        'C': {'D': 1},
        'D': {}
    }
    distances_negative = bellman_ford(graph_negative, 'A')
    assert distances_negative == {'A': 0, 'B': 1, 'C': -2, 'D': -1}, "❌ Test échoué: distances incorrectes avec arêtes négatives"
    print("✅ Test 2 réussi")

    # Test 3: Vérifier le comportement avec un graphe déconnecté
    graph_disconnected = {
        'A': {'B': 1},
        'B': {},
        'C': {'D': 2},
        'D': {}
    }
    distances_disconnected = bellman_ford(graph_disconnected, 'A')
    assert distances_disconnected == {'A': 0, 'B': 1, 'C': float('inf'), 'D': float('inf')}, "❌ Test échoué: distances incorrectes pour graphe déconnecté"
    print("✅ Test 3 réussi")
    
test_bellman_ford()

File: algorithms/graphs/bellman_ford/typescript/bellman_ford.ts
/*
-----------------------------------------------------------------------------
Nom du fichier : bellman_ford.ts
Auteur        : CAPELLE Gabin (https://github.com/CapelleGab)
Créé le      : 10-06-2025 20:39
Dernière MAJ : 10-06-2025 20:39

Description   : Implémentation de l'algorithme de Bellman-Ford en TypeScript.
-----------------------------------------------------------------------------
*/

const bellman_ford = (Graph: any, start: string) => {
  const distances = {};
  const previous = {};

  for (const vertex of Graph.vertices) {
    distances[vertex] = Infinity;
    previous[vertex] = null;
  }

  distances[start] = 0;

  for (let i = 0; i < Graph.vertices.length - 1; i++) {
    for (const vortex of Graph.vertices) {
      for (const neighbor of Object.keys(Graph.adjacencyList[vortex])) {
        const distance =
          distances[vortex] + Graph.adjacencyList[vortex][neighbor];
        if (distance < distances[neighbor]) {
          distances[neighbor] = distance;
          previous[neighbor] = vortex;
        }
      }
    }
  }
  return { distances, previous };
};

const test_bellman_ford = () => {
  const graph = {
    vertices: ["A", "B", "C", "D"],
    adjacencyList: {
      A: { B: 1, C: 4 },
      B: { C: 2, D: 5 },
      C: { D: 1 },
      D: {},
    },
  };
  const result = bellman_ford(graph, "A");
  console.log(result);

  const graph_negative = {
    vertices: ["A", "B", "C", "D"],
    adjacencyList: {
      A: { B: 1, C: 4 },
      B: { C: -3, D: 2 },
      C: { D: 1 },
      D: {},
    },
  };
  const result_negative = bellman_ford(graph_negative, "A");
  console.log(result_negative);

  const graph_disconnected = {
    vertices: ["A", "B", "C", "D"],
    adjacencyList: {
      A: { B: 1 },
      B: { C: 2 },
      C: { D: 1 },
      D: {},
    },
  };
  const result_disconnected = bellman_ford(graph_disconnected, "A");
  console.log(result_disconnected);
};

test_bellman_ford();
